### PR TITLE
options.isMergeableObject = function(value, optionsArgument, key)

### PR DESCRIPTION
**Pull request by bluelovers**
```ts
options.isMergeableObject = function(value, optionsArgument, key)
```
allow user plugin able

### CODE REVIEW

**Comment by TehShrike** [comment]
I'm not opposed to letting users pass in a custom `isMergeableObject` function.  With some caveats:

1. it shouldn't take `key`/`index` as an argument - that shouldn't have anything to do with object mergeability.
2. we definitely need some tests first
3. the `isMergeableObject` function that the user passes in should take two arguments: the value to be tested, and the original internal `isMergeableObject` function, so that an implementation could look like this:

```js
function customIsMergeableObject(value, isMergeable) { 
  return value && (value[Symbol.iterator] || isMergeable(value)
}
```
(or whatever.)

This would allow people to easily add on to the definition of what was mergeable if they didn't want to override it completely.

I don't think custom mergeable functions need the entire options object, though.

**Comment by bluelovers** [comment]
something about moment object need this

File: index.js [hunk 1]
```diff
@@ -4,30 +4,41 @@ function emptyTarget(val) {
...
 		: value
 }
 
+function _isMergeableObject(value, optionsArgument, key) {
+	let ret;
+	if (optionsArgument && optionsArgument.isMergeableObject) {
+		ret = optionsArgument.isMergeableObject(value, isMergeableObject, optionsArgument, key)
+	}
+	if (ret === null || typeof ret === 'undefined') {
+		ret = isMergeableObject(value)
```
Comment (TehShrike) [inline index.js]: This seems like a huge issue - we shouldn't fall back to the default function.

This function seems like it should just be something like

```js
function isMergeableObject(value, optionsArgument) {
	const testFunction = (optionsArgument && optionsArgument.isMergeableObject) 
		? optionsArgument.isMergeableObject
		: defaultIsMergeableObject

	return testFunction(value, defaultIsMergeableObject)
}
```

**Comment by bluelovers** [comment]
```ts
isMergeableObject(value, isMergeable)
			{
				let bool = isMergeable(value);

				if (bool && typeof value == 'object' && !Array.isArray(value))
				{
					bool = isPlainObject(value);
				}

				return bool;
			}
```

**Pull request closed without merging**